addons/token-exchange/token_exchanger_addon.go

```diff
@@ -0,0 +1,199 @@
+	"manifests/spoke_deployment.yaml",
+}
+
+var agentHubPermissionFiles = []string{
+	"manifests/hub_role.yaml",
+	"manifests/hub_rolebinding.yaml",
+	"manifests/hub_clusterrole.yaml",
+	"manifests/hub_clusterrolebinding.yaml",
+}
+
+//go:embed manifests
+var manifestFiles embed.FS
+
+type TokenExchangeAddon struct {
+	KubeClient kubernetes.Interface
+	Recorder   events.Recorder
+	AgentImage string
+}
+
+func init() {
+	scheme.AddToScheme(genericScheme)
+}
+
+// Manifests generates manifestworks to deploy the token exchange addon agent on the managed cluster
+func (a *TokenExchangeAddon) Manifests(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+	objects := []runtime.Object{}
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = "default"
+	}
+
+	if len(a.AgentImage) == 0 {
+		return objects, fmt.Errorf("image not provided for agent %q", TokenExchangeName)
+	}
+
+	manifestConfig := struct {
+		KubeConfigSecret      string
+		ClusterName           string
+		AddonInstallNamespace string
+		Image                 string
+	}{
+		KubeConfigSecret:      fmt.Sprintf("%s-hub-kubeconfig", TokenExchangeName),
+		AddonInstallNamespace: installNamespace,
+		ClusterName:           cluster.Name,
+		Image:                 a.AgentImage,
+	}
+
+	for _, file := range agentDeploymentFiles {
+		template, err := manifestFiles.ReadFile(file)
+		if err != nil {
+			return objects, err
+		}
+		raw := assets.MustCreateAssetFromTemplate(file, template, &manifestConfig).Data
+		object, _, err := genericCodec.Decode(raw, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, object)
+	}
+
+	return objects, nil
+}
+
+// GetAgentAddonOptions returns the options of token exchange addon agent
+func (a *TokenExchangeAddon) GetAgentAddonOptions() agent.AgentAddonOptions {
+	return agent.AgentAddonOptions{
+		AddonName: TokenExchangeName,
+		Registration: &agent.RegistrationOption{
+			CSRConfigurations: agent.KubeClientSignerConfigurations(TokenExchangeName, TokenExchangeName),
+			CSRApproveCheck:   a.csrApproveCheck,
+			PermissionConfig:  a.permissionConfig,
+		},
+	}
+}
+
+// To check the addon agent csr, we check
+// 1. if the signer name in csr request is valid.
+// 2. if organization field and commonName field in csr request is valid.
+// 3. if user name in csr is the same as commonName field in csr request.
+func (a *TokenExchangeAddon) csrApproveCheck(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn, csr *certificatesv1.CertificateSigningRequest) bool {
+	groups := agent.DefaultGroups(cluster.Name, TokenExchangeName)
+	clusterAddOnGroup := groups[0]
+	addOnGroup := groups[1]
+	authenticatedGroup := groups[2]
+	agentUserName := agent.DefaultUser(cluster.Name, TokenExchangeName, TokenExchangeName)
+
+	if csr.Spec.SignerName != certificatesv1.KubeAPIServerClientSignerName {
+		return false
```

Comment:
no error log?

Reply:
Not sure if the logs would be helpful here.
request comes from an agent running on different cluster and may not see these logs.
adding error logs here would confuse the user on this cluster. At best, it could be info logs for debug.

`token_exchanger_addon.go` runs on one cluster and `token_exchanger_agent.go` runs on another cluster.

Reply:
I see, maybe Debug level then? While debugging it's useful to have as much info as possible to really understand why something was skipped.